zerver/lib/push_notifications.py

```diff
@@ -986,6 +1010,18 @@ def get_message_payload_gcm(
     data = get_message_payload(
         user_profile, message, mentioned_user_group_id, mentioned_user_group_name
     )
+
+    if message.is_stream_message() and not check_can_access_user(message.sender, user_profile):
```

Comment:
We should avoid doing this check in several places -- it's expensive, so doing 3-4 for a given push notifications API call can be expensive. It shouldn't be too hard to just pass a `can_access_sender` boolean down from an outer function.